sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/Samplers.java

```diff
@@ -231,19 +259,109 @@ public SamplingResult shouldSample(
         Kind spanKind,
         ReadableAttributes attributes,
         List<Link> parentLinks) {
-      if (parentContext.isValid()) {
-        if (parentContext.getTraceFlags().isSampled()) {
-          return EMPTY_RECORDED_AND_SAMPLED_SAMPLING_RESULT;
-        }
-        return EMPTY_NOT_SAMPLED_OR_RECORDED_SAMPLING_RESULT;
+      if (!parentContext.isValid()) {
+        return this.root.shouldSample(
+            parentContext, traceId, name, spanKind, attributes, parentLinks);
       }
-      return this.delegateSampler.shouldSample(
-          parentContext, traceId, name, spanKind, attributes, parentLinks);
+
+      if (parentContext.isRemote()) {
+        return parentContext.getTraceFlags().isSampled()
+            ? this.remoteParentSampled.shouldSample(
+                parentContext, traceId, name, spanKind, attributes, parentLinks)
+            : this.remoteParentNotSampled.shouldSample(
+                parentContext, traceId, name, spanKind, attributes, parentLinks);
+      }
+      return parentContext.getTraceFlags().isSampled()
+          ? this.localParentSampled.shouldSample(
+              parentContext, traceId, name, spanKind, attributes, parentLinks)
+          : this.localParentNotSampled.shouldSample(
+              parentContext, traceId, name, spanKind, attributes, parentLinks);
     }
 
     @Override
     public String getDescription() {
-      return String.format("ParentOrElse{%s}", this.delegateSampler.getDescription());
+      return String.format(
+          "ParentBased{root:%s,remoteParentSampled:%s,remoteParentNotSampled:%s,"
+              + "localParentSampled:%s,localParentNotSampled:%s}",
+          this.root.getDescription(),
+          this.remoteParentSampled.getDescription(),
+          this.remoteParentNotSampled.getDescription(),
+          this.localParentSampled.getDescription(),
+          this.localParentNotSampled.getDescription());
+    }
+
+    static class Builder {
+      private final Sampler root;
+      private Sampler remoteParentSampled;
+      private Sampler remoteParentNotSampled;
+      private Sampler localParentSampled;
+      private Sampler localParentNotSampled;
+
+      /**
+       * Sets the remoteParentSampled. Optional.
+       *
+       * @param remoteParentSampled used to make the sampling decisions if the remote parent exist
+       *     and is sampled.
+       * @return this.
+       */
+      public Builder withRemoteParentSampled(Sampler remoteParentSampled) {
```

Comment:
Our other builders like `SpanBuilder` and `Attributes.Builder` use `set` instead of `with`, can you use set here too?

Reply:
Alternatively, `when` seems like it could fit as a special semantics for this class, in particular I think if the variable was named `Sampler whenRemoteParentSampled` it's really easy to read.

Reply:
> named Sampler whenRemoteParentSampled it's really easy to read.
+1

Reply:
I think I'd go with `set` for consistency with the other APIs.

Reply:
thanks. Let's use `set`  instead of `with`.